### PR TITLE
PTX-12097 Validate new Telemetry method and regression test changes

### PR DESCRIPTION
**Pull request by nikolaypopov**
Validate new Telemetry method and regression test changes

Signed-off-by: nikolaypopov <nikolay.popov86@gmail.com>

### CODE REVIEW

File: test/integration_test/basic_test.go [hunk 1]
```diff
@@ -439,67 +416,56 @@ func BasicUpgradeOperator(tc *types.TestCase) func(*testing.T) {
...
-}
+		// Create StorageCluster
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+
+		// Validate Telemetry is not enabled by default
+		logrus.Info("Validate Telemetry is not enabled by default")
+		if cluster.Spec.Monitoring != nil {
+			if cluster.Spec.Monitoring.Telemetry != nil {
+				if cluster.Spec.Monitoring.Telemetry.Enabled {
+					require.False(t, cluster.Spec.Security.Enabled, "failed to validate default Telemetry status: expected: false, actual: %v", cluster.Spec.Monitoring.Telemetry.Enabled)
```
Comment (pure-jliao) [inline test/integration_test/basic_test.go:439]: nit: by default the field is nil, we expect it to be nil or false, so if it's enabled, we can call `require.FailNow()`. but this should work fine.

File: pkg/util/test/util.go [hunk 2]
```diff
@@ -776,6 +781,12 @@ func ValidateUninstallStorageCluster(
 		return err
 	}
 
+	// Verify telemetry secret is deleted on UninstallAndWipe
```
Comment (pure-jliao) [inline pkg/util/test/util.go:784]: can we check the delete strategy to be UninstallAndWipe first then validate the secret deletion?

File: pkg/util/test/util.go [hunk 3]
```diff
@@ -2337,13 +2351,31 @@ func ValidateTelemetryUninstalled(pxImageList map[string]string, cluster *corev1
...
+		return err
+	}
+	logrus.Infof("PX Operator version: %v", opVersion)
+
 	if cluster.Spec.Monitoring != nil &&
 		cluster.Spec.Monitoring.Telemetry != nil &&
 		cluster.Spec.Monitoring.Telemetry.Enabled {
+		logrus.Info("Telemetry is enabled in StorageCluster")
+		if pxVersion.GreaterThanOrEqual(pxVer2_12) && pxVersion.GreaterThanOrEqual(opVer1_9_1) {
+			return ValidateTelemetryEnabled(pxImageList, cluster, timeout, interval)
```
Comment (pure-jliao) [inline pkg/util/test/util.go:2369]: are you implementing a new validation function for ccm-go only? that case can we call it `ValidateTelemetryV2Installed` and change the old one to `ValidateTelemetryV1Installed` to be more clear?

File: pkg/util/test/util.go [hunk 4]
```diff
@@ -2337,13 +2351,31 @@ func ValidateTelemetryUninstalled(pxImageList map[string]string, cluster *corev1
...
+		}
 		return ValidateTelemetryInstalled(pxImageList, cluster, timeout, interval)
 	}
 
-	return ValidateTelemetryUninstalled(pxImageList, cluster, timeout, interval)
+	logrus.Info("Telemetry is disabled in StorageCluster")
+	if pxVersion.GreaterThanOrEqual(pxVer2_12) && pxVersion.GreaterThanOrEqual(opVer1_9_1) {
+		return ValidateTelemetryDisabled(cluster, timeout, interval)
```
Comment (pure-jliao) [inline pkg/util/test/util.go:2376]: same here.

File: pkg/util/test/util.go [hunk 5]
```diff
@@ -2578,6 +2610,255 @@ func ValidateAlertManagerDisabled(pxImageList map[string]string, cluster *corev1
...
+		return fmt.Errorf("failed to find image for envoy")
+	}
+	envoyImageName = util.GetImageURN(cluster, envoyImageName)
+	if registrationServiceDep.Spec.Template.Spec.Containers[1].Image != envoyImageName {
+		return fmt.Errorf("envoy image mismatch, image: %s, expected: %s",
+			registrationServiceDep.Spec.Template.Spec.Containers[1].Image,
+			envoyImageName)
+	}
+
+	// Verify px-telemetry-metrics-collector deployment and images
```
Comment (pure-jliao) [inline pkg/util/test/util.go:2655]: can we validate the deployment has a init container as well?

File: pkg/util/test/util.go [hunk 6]
```diff
@@ -2578,6 +2610,255 @@ func ValidateAlertManagerDisabled(pxImageList map[string]string, cluster *corev1
...
+	logrus.Info("Validate px-telemetry-phonehome daemonset and images")
+	if err := appops.Instance().ValidateDaemonSet("px-telemetry-phonehome", cluster.Namespace, timeout); err != nil {
+		return err
+	}
+	telemetryPhonehomeDs, err := appops.Instance().GetDaemonSet("px-telemetry-phonehome", cluster.Namespace)
+	if err != nil {
+		return err
+	}
+
+	// Verify collector container image inside px-telemetry-phonehome
```
Comment (pure-jliao) [inline pkg/util/test/util.go:2699]: same here, the template should have an init container, we can just check the image.

File: pkg/util/test/util.go [hunk 7]
```diff
@@ -2578,6 +2610,255 @@ func ValidateAlertManagerDisabled(pxImageList map[string]string, cluster *corev1
...
+	if err != nil {
+		return err
+	}
+
+	// Verify telemetry serviceaccounts
+	_, err = coreops.Instance().GetServiceAccount("px-telemetry", cluster.Namespace)
+	if err != nil {
+		return err
+	}
+
```
Comment (pure-jliao) [inline pkg/util/test/util.go:2774]: are we able to validate the clusterrole and clusterrolebinding as well?

File: pkg/util/test/util.go [hunk 8]
```diff
@@ -2578,6 +2610,255 @@ func ValidateAlertManagerDisabled(pxImageList map[string]string, cluster *corev1
...
+		if !errors.IsNotFound(err) {
+			return "", true, fmt.Errorf("found px-telemetry-tls-certificate configmap, waiting for deletion")
+		}
+
+		// Verify telemetry serviceaccounts
+		_, err = coreops.Instance().GetServiceAccount("px-telemetry", cluster.Namespace)
+		if !errors.IsNotFound(err) {
+			return "", true, fmt.Errorf("found px-telemetry serviceaccount, waiting for deletion")
+		}
+
```
Comment (pure-jliao) [inline pkg/util/test/util.go:2850]: same here for clusterrole and clusterrolebinding

**Review by pure-jliao — APPROVED** [review]
LGTM after addressing comments.

File: test/integration_test/basic_test.go (hunk 1 above)
Comment (nikolaypopov) [reply test/integration_test/basic_test.go:439]: If this is fine I will just leave it as is

File: pkg/util/test/util.go (hunk 2 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go:784]: In this func we only validate after the wipe, cluster and all components are gone at this point, the func that uninstalls cluster is setting the UninstallAndWipe strategy

File: pkg/util/test/util.go (hunk 3 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go]: done

File: pkg/util/test/util.go (hunk 4 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Done

File: pkg/util/test/util.go (hunk 5 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go:2657]: Done

File: pkg/util/test/util.go (hunk 6 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go:2716]: Done

File: pkg/util/test/util.go (hunk 7 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go:2803]: Done

File: pkg/util/test/util.go (hunk 8 above)
Comment (nikolaypopov) [reply pkg/util/test/util.go:2892]: Done

**Pull request merged**